crates/project/src/search_history.rs

```diff
@@ -0,0 +1,262 @@
+use collections::HashMap;
+use smallvec::SmallVec;
+const SEARCH_HISTORY_LIMIT: usize = 20;
+
+/// Determines the behavior to use when inserting a new query into the search history.
+#[derive(Default, Debug, Clone, PartialEq)]
+pub enum QueryInsertionBehavior {
+    #[default]
+    /// Always insert the query to the search history.
+    AlwaysInsert,
+    /// Replace the previous query in the search history, if the new query contains the previous query.
+    ReplacePreviousIfContains,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct SearchHistorySelectionHandle(usize);
+
+#[derive(Default, Debug, Clone)]
+pub struct SearchHistory {
+    history: SmallVec<[String; SEARCH_HISTORY_LIMIT]>,
+    selected: HashMap<SearchHistorySelectionHandle, Option<usize>>,
+    insertion_behavior: QueryInsertionBehavior,
+}
+
+impl SearchHistory {
+    pub fn new_with_handle(
```

Comment:
NIT: feels simpler to do another call to `new_handle` method below than write `let (a, b) = ` but not insisting.